arch/arm/core/cortex_m/mpu/arm_mpu_v8_internal.h

```diff
@@ -0,0 +1,139 @@
+/*
+ * Copyright (c) 2017 Linaro Limited.
+ * Copyright (c) 2018 Nordic Semiconductor ASA.
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <cmse.h>
+
+/* Global MPU configuration at system initialization. */
+static void _mpu_init(void)
+{
+	/* Configure the cache-ability attributes for all the
+	 * different types of memory regions.
+	 */
+
+	/* Flash region(s): Attribute-0
+	 * SRAM region(s): Attribute-1
+	 */
+	MPU->MAIR0 =
+		((MPU_MAIR_ATTR_FLASH & MPU_MAIR0_Attr0_Msk) >>
+			MPU_MAIR0_Attr0_Pos)
+		|
+		((MPU_MAIR_ATTR_SRAM & MPU_MAIR0_Attr1_Msk) >>
+			MPU_MAIR0_Attr1_Pos);
+}
+
+/* This internal function performs MPU region initialization.
+ *
+ * Note:
+ *   The caller must provide a valid region index.
+ */
+static void _region_init(u32_t index, struct arm_mpu_region *region_conf)
+{
+	ARM_MPU_SetRegion(
+		/* RNR */
+		index,
+		/* RBAR */
+		(region_conf->base & MPU_RBAR_BASE_Msk)
+		| (region_conf->attr.rbar &
+			(MPU_RBAR_XN_Msk | MPU_RBAR_AP_Msk | MPU_RBAR_SH_Msk)),
+		/* RLAR */
+		(region_conf->attr.r_limit & MPU_RLAR_LIMIT_Msk)
+		| ((region_conf->attr.mair_idx << MPU_RLAR_AttrIndx_Pos)
+			& MPU_RLAR_AttrIndx_Msk)
+		| MPU_RLAR_EN_Msk
+	);
+
+	SYS_LOG_DBG("[%d] 0x%08x 0x%08x 0x%08x 0x%08x",
+			index, region_conf->base, region_conf->attr.rbar,
+			region_conf->attr.mair_idx, region_conf->attr.r_limit);
+}
+
+#if defined(CONFIG_USERSPACE) || defined(CONFIG_MPU_STACK_GUARD) || \
+	defined(CONFIG_APPLICATION_MEMORY)
+
+/**
+ * This internal function allocates default RAM cache-ability, share-ability,
+ * and execution allowance attributes along with the requested access
+ * permissions and size.
+ */
+static inline void _get_mpu_ram_region_attr(arm_mpu_region_attr_t *p_attr,
+	u32_t ap, u32_t base, u32_t size)
+{
+	p_attr->rbar = ((1 << MPU_RBAR_XN_Pos) & MPU_RBAR_XN_Msk)
+		| ((ap << MPU_RBAR_AP_Pos) & MPU_RBAR_AP_Msk);
+	p_attr->mair_idx = MPU_MAIR_INDEX_SRAM;
+	p_attr->r_limit = REGION_LIMIT_ADDR(base, size);
+}
+
+/**
+ * This internal function is utilized by the MPU driver to combine a given
+ * MPU attribute configuration and region size and fill-in a structure with
+ * the correct parameter set.
+ */
+static inline void _get_region_attr_by_conf(arm_mpu_region_attr_t *p_attr,
+	u32_t ap_attr, u32_t base, u32_t size)
+{
+	p_attr->rbar = ap_attr  & (MPU_RBAR_XN_Msk | MPU_RBAR_AP_Msk);
+	p_attr->mair_idx = MPU_MAIR_INDEX_SRAM;
```

Comment:
Should we unconditionally overwrite mair_idx here?

Reply:
Good point.
The function is used to re-configure RAM regions so this is safe to do.
I can document more clearly that _get_region_attr_by_conf() is used for RAM regions, if you suggest